libbpf-tools/syscount.c

```diff
@@ -0,0 +1,1167 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Anton Protopopov
+//
+// Based on syscount(8) from BCC by Sasha Goldshtein
+#include <sys/resource.h>
+#include <unistd.h>
+#include <signal.h>
+#include <fcntl.h>
+#include <time.h>
+#include <argp.h>
+#include <bpf/bpf.h>
+#include "syscount.h"
+#include "syscount.skel.h"
+
+#define MAX(x, y) (((x) > (y)) ? (x) : (y))
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+const char *argp_program_version = "syscount 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+static const char argp_program_doc[] =
+"\nsyscount: summarize syscall counts and latencies\n"
+"\n"
+"EXAMPLES:\n"
+"    syscount                 # print top 10 syscalls by count every second\n"
+"    syscount -p $(pidof dd)  # look only at a particular process\n"
+"    syscount -L              # measure and sort output by latency\n"
+"    syscount -P              # group statistics by pid, not by syscall\n"
+"    syscount -x -i 5         # count only failed syscalls\n"
+"    syscount -e ENOENT -i 5  # count only syscalls failed with a given errno"
+;
+
+static const struct argp_option opts[] = {
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ "pid", 'p', "PID", 0, "Process PID to trace" },
+	{ "interval", 'i', "INTERVAL", 0, "Print summary at this interval"
+				" (seconds), 0 for infinite wait (default)" },
+	{ "duration", 'd', "DURATION", 0, "Total tracing duration (seconds)" },
+	{ "top", 'T', "TOP", 0, "Print only the top syscalls (default 10)" },
+	{ "failures", 'x', NULL, 0, "Trace only failed syscalls" },
+	{ "latency", 'L', NULL, 0, "Collect syscall latency" },
+	{ "milliseconds", 'm', NULL, 0, "Display latency in milliseconds"
+					" (default: microseconds)" },
+	{ "process", 'P', NULL, 0, "Count by process and not by syscall" },
+	{ "errno", 'e', "ERRNO", 0, "Trace only syscalls that return this error"
+				 "(numeric or EPERM, etc.)" },
+	{ "list", 'l', NULL, 0, "Print list of recognized syscalls and exit" },
+	{},
+};
+
+static struct env {
+	bool list_syscalls;
+	bool milliseconds;
+	bool failures;
+	bool verbose;
+	bool latency;
+	bool process;
+	int filter_errno;
+	int interval;
+	int duration;
+	int top;
+	pid_t pid;
+} env = {
+	.top = 10,
+};
+
+static int get_int(const char *arg, int *ret, int min, int max)
+{
+	char *end;
+	int val;
+
+	errno = 0;
+	val = strtol(arg, &end, 10);
+	if (errno) {
+		warn("strtol: %s: %s\n", arg, strerror(errno));
+		return -1;
+	} else if (end == arg || val < min || val > max) {
```

Comment:
wouldn't `*end == 0` be a better check that entire string argument was a valid integer, parsed completely?

Reply:
Well, for example, here https://github.com/iovisor/bcc/pull/2920/files/3c40184699520362fece1d9ae3fe23d6182b9520#diff-68934a0eef57fae11fc290164fd99838R288 this is not the case. When I move the errno/syscall parsing to a separate file, this, probably, will. However, I also wanted to distinguish cases when errno is set vs. no symbols parsed. I will take a better look in the v2.

Reply:
link doesn't work, but sure, do whatever makes sense, I guess